### ISSUE TITLE
CDC lag up to 1.5h when there are 100 changefeeds and CDC rolling restart

### ISSUE THREAD

**Issue by fubinzh** [pingcap/tiflow#10797]
### What did you do?

1. TiDB cluster deployed in GCP GKE env, with 24 TiKV (16c64g) and 9 CDC node (16c64g), cluster size ~40TB. 3 workload running, one workload with row width ~1mb, one 9kb, one 1.7kb.
2. 100 changefeed created, each changefeed cover 40 tables.
3. update the cdc configuration to trigger rolling restart

### What did you expect to see?

CDC lag should be less than 10s during rolling restart

### What did you see instead?

CDC lag up to 1.5h
![image](https://github.com/pingcap/tiflow/assets/7403864/b004c5ca-4729-4536-8c59-345ba68c86b1)

### Versions of the cluster

cdc version:
Release Version: v8.0.0                                  
Git Commit Hash: 130403f3b9b8ad8a28ceada642277986e317ebc2
Git Branch: heads/refs/tags/v8.0.0                       
UTC Build Time: 2024-03-15 13:58:58                      
Go Version: go version go1.21.6 linux/amd64              
Failpoint Build: false

**Comment by sdojjy** [comment]
there are 4000 tables in one single database,  every changefeed needs to load all of them to initialize the schema storage(costs about 7 seconds).  
the owner and processor `tick` loop will hang for a long time if there are lots of changefeeds.
```
encoding/json.Unmarshal({0xc0501ea000, 0xfafc, 0x10000}, {0x4e344a0, 0xc025eca540})
        encoding/json/decode.go:108 +0x111
github.com/pingcap/tidb/pkg/meta.(*Meta).ListTables(0xc0353ba000, 0xaa)
        github.com/pingcap/tidb@v1.1.0-beta.0.20240228123331-27ce02afd2e3/pkg/meta/meta.go:1002 +0x2a9
github.com/pingcap/tiflow/cdc/entry/schema.NewSnapshotFromMeta({{0xc0033a5a82, 0x7}, {0xc0033a5a9a, 0x6}}, 0x4b436c0?, 0x63946f5cab8019f, 0x0, {0x5c1a088, 0xc0350fa1c0})
        github.com/pingcap/tiflow/cdc/entry/schema/snapshot.go:171 +0x3ac
github.com/pingcap/tiflow/cdc/entry.NewSchemaStorage({0x5c512b8?, 0xc00172f140?}, 0x63946f5cab8019f, 0x0, {{0xc0033a5a82?, 0xc0038feb20?}, {0xc0033a5a9a?, 0x38?}}, 0x2, {0x5c1a088, ...})
        github.com/pingcap/tiflow/cdc/entry/schema_storage.go:103 +0xbc
github.com/pingcap/tiflow/cdc/processor.(*processor).initDDLHandler(0xc003fe5880, {0x7f76c51a4948, 0xc03c8cbae8})
        github.com/pingcap/tiflow/cdc/processor/processor.go:720 +0x17f
github.com/pingcap/tiflow/cdc/processor.(*processor).lazyInitImpl(0xc003fe5880, {0x5c24f40, 0xc03c8cba10})
        github.com/pingcap/tiflow/cdc/processor/processor.go:593 +0x2d6
github.com/pingcap/tiflow/cdc/processor.(*processor).tick(0xc003fe5880, {0x5c24f40?, 0xc03c8cba10})
        github.com/pingcap/tiflow/cdc/processor/processor.go:553 +0x4f
github.com/pingcap/tiflow/cdc/processor.(*processor).Tick(0xc003fe5880, {0x5c24f40, 0xc03c8cba10}, 0xa39bc40?, 0xc1762e1a23ff35cc?)
```

**Comment by flowbehappy** [comment]
Addressed by this issue https://github.com/pingcap/tiflow/issues/10845

**Comment by sdojjy** [comment]
Root cause:
1. ticdc synchronously initializes all changefeeds within a tick call.
2. During the initialization of changefeeds, it needs to fetch schema information related to the changefeed's tables. However, because TiDB's schema API does not support filtering, ticdc needs to retrieve all schema information from TiDB and locally filter out unnecessary tables. This step is very time-consuming; in practical testing, it takes around 8 seconds to process 4000 tables. In this particular test scenario, with 100 changefeeds, each requiring 8 seconds for initialization, the total time will be around 800 seconds.

**Comment by asddongmen** [comment]
After [#10832](https://github.com/pingcap/tiflow/pull/10832) and [#11095](https://github.com/pingcap/tiflow/pull/11095) merged, the lag in changefeeds dropped from 1.5 hours to 51 seconds during a CDC rolling restart. (With 5 cdc servers.)

**Comment by flowbehappy** [comment]
This issue is resolved by the PRs of https://github.com/pingcap/tiflow/issues/11124 and https://github.com/pingcap/tiflow/issues/10845. And those issue are all addressed with PR merged to corresponding LTS(v7.5.2 and v8.1.0). So I will close this issue and remove the affect version to avoid the duplicate in issue tracking system.
